ddtrace/tracer/textmap.go

```diff
@@ -487,3 +501,292 @@ func (*propagatorB3) extractTextMap(reader TextMapReader) (ddtrace.SpanContext,
+			}
+			parentHeader = v
+		case tracestateHeader:
+			stateHeader = v
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	var ctx spanContext
+	if err := parseTraceparent(&ctx, parentHeader); err != nil {
+		return nil, err
+	}
+	if err := parseTracestate(&ctx, stateHeader); err != nil {
+		return nil, err
+	}
+	// ctx.traceID might be nil due to higher order bits, eg. 1....00000000
+	if ctx.spanID == 0 {
+		return nil, ErrSpanContextNotFound
+	}
+	return &ctx, nil
+}
+
+// parseTraceparent attempts to parse traceparentHeader which describes the position
+// of the incoming request in its trace graph in a portable, fixed-length format.
+// The format of the traceparentHeader is `-` separated string with in the
+// following format: `version-traceId-spanID-flags`,
+// where:
+// - version - represents the version of the W3C Tracecontext Propagation format in hex format.
+// - traceId - represents the propagated traceID in the format of 32 hex-encoded digits.
+// - spanID - represents the propagated spanID in the format of 16 hex-encoded digits.
+// - flags - represents the propagated flags in the format of 2 hex-encoded digits, and supports 8 unique flags.
+// Example value of HTTP `traceparent` header: `00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01`,
+// Currently, Go tracer doesn't support 128-bit traceIDs, so the full traceID (32 hex-encoded digits) must be
+// stored into a field that is accessible from the span’s context. TraceId will be parsed from the least significant 16
+// hex-encoded digits into a 64-bit number.
+func parseTraceparent(ctx *spanContext, header string) error {
+	nonWordRegex := "\\W_"
+	header = strings.ToLower(strings.Trim(header, "\t -"))
+	if len(header) == 0 {
+		return ErrSpanContextNotFound
+	}
+	if len(header) != 55 {
+		return ErrSpanContextCorrupted
+	}
+	parts := strings.Split(header, "-")
+	if len(parts) != 4 {
+		return ErrSpanContextCorrupted
+	}
+	version := strings.Trim(parts[0], nonWordRegex)
+	if len(version) != 2 {
+		return ErrSpanContextCorrupted
+	}
+	if v, err := strconv.ParseUint(version, 16, 64); err != nil || v == 255 {
+		return ErrSpanContextCorrupted
+	}
+	// parsing traceID
+	fullTraceID := strings.Trim(parts[1], nonWordRegex)
+	if len(fullTraceID) != 32 {
+		return ErrSpanContextCorrupted
+	}
+	// checking that the entire TraceID is a valid hex string
+	if ok, err := regexp.MatchString("^[a-f0-9]+$", fullTraceID); !ok || err != nil {
+		return ErrSpanContextCorrupted
+	}
+	var err error
+	if ctx.traceID, err = strconv.ParseUint(fullTraceID[16:], 16, 64); err != nil {
+		return ErrSpanContextCorrupted
+	}
+	if ctx.traceID == 0 {
+		if strings.Trim(fullTraceID[:16], "0") == "" {
+			return ErrSpanContextNotFound
+		}
+	}
+	// parsing spanID
+	spanID := strings.Trim(parts[2], nonWordRegex)
+	if len(spanID) != 16 {
+		return ErrSpanContextCorrupted
+	}
+	if ok, err := regexp.MatchString("[a-f0-9]+", spanID); !ok || err != nil {
+		return ErrSpanContextCorrupted
+	}
```

Comment:
Do we need this regex? Unlike the traceID, we're parsing the full spanID, so I think ParseUint should tell us if it's valid or not, no?

If we *do* need it, it should probably be an exact match (with the `^` and `$`) like for the `fullTraceID` above.

Reply:
`-34567899`, `.3456789` would be parsed when they shouldn't. 
WIll add the `^` and `$`